### DISCUSSION TITLE
Wildcard Variables

### DISCUSSION THREAD

**Discussion by WASasquatch** [adieyal/sd-dynamic-prompts#303] [Ideas]
A wildcard being used, can be a name, and reused elsewhere in the prompt, or negative prompt. 

For example take the NSP wildcard `__animals__` one could assign a name, maybe like `__animals:ani1__`  which parses to `dog` *(or whatever random animal, just an example)*

Then, a user can use that parsed wildcard anywhere else, say the negative prompt: `((__ani1__ face))` which would be `((dog face))`

Example usages:

- Manipulating subjects with other subjects, but not wanting, say a dog face, but other features.
- Wanting a certain type of object, but not other types in a similar name/style.
- Etc

_PS I'll look at source, see if this is something I can PR but pretty busy._

**Comment by adieyal** [comment]
Have you had a look at [Jinja templates](https://github.com/adieyal/sd-dynamic-prompts/blob/main/jinja2.md)? Not exactly what you want, but might take you part of the way.

**Reply by WASasquatch** [reply]
> Have you had a look at [Jinja templates](https://github.com/adieyal/sd-dynamic-prompts/blob/main/jinja2.md)? Not exactly what you want, but might take you part of the way.

That's pretty cool for programmatically doing prompts, I guess, though at that point you just have WebUI scripts too with Python directly. Jinja interpreter to Python is pretty slow for something so simple with prompts. Lots of overhead.

As for what I'm looking for, it would be extreme overkill just to store the result of a wildcard *already being handled* by sd-dynamic-prompts. 

Imo, to prevent all the quirks of WebUI, and all the errors, etc, we should be trying to limit back-and-fourth injection type hijacking of data by other libraries/apis.

**Comment by akx** [comment]
Variables were implemented by yours truly in https://github.com/adieyal/dynamicprompts/pull/63.

* See https://github.com/adieyal/sd-dynamic-prompts/blob/main/docs/SYNTAX.md#variables for the syntax.
* You'll need dynamicprompts >= [v0.21.0](https://github.com/adieyal/dynamicprompts/releases/tag/v0.21.0) for this to work.
